ql/src/java/org/apache/hadoop/hive/ql/exec/tez/monitoring/TezProgressMonitor.java

```diff
@@ -0,0 +1,315 @@
+package org.apache.hadoop.hive.ql.exec.tez.monitoring;
+
+import org.apache.hadoop.hive.ql.log.ProgressMonitor;
+import org.apache.hadoop.hive.ql.plan.BaseWork;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.tez.dag.api.TezException;
+import org.apache.tez.dag.api.client.DAGClient;
+import org.apache.tez.dag.api.client.DAGStatus;
+import org.apache.tez.dag.api.client.Progress;
+import org.apache.tez.dag.api.client.StatusGetOpts;
+import org.apache.tez.dag.api.client.VertexStatus;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import static org.apache.tez.dag.api.client.DAGStatus.State.KILLED;
+
+class TezProgressMonitor implements ProgressMonitor {
+  private static final int COLUMN_1_WIDTH = 16;
+  private final Map<String, BaseWork> workMap;
+  private final SessionState.LogHelper console;
+  private final long executionStartTime;
+  private final DAGStatus status;
+  Map<String, VertexStatus> vertexStatusMap = new HashMap<>();
+  Map<String, StateFullProgress> progressCountsMap = new HashMap<>();
+
+  /**
+   * Try to get most the data required from dagClient in the constructor itself so that even after
+   * the tez job has finished this object can be used for later use.s
+   */
+  TezProgressMonitor(DAGClient dagClient, Map<String, BaseWork> workMap,
+      SessionState.LogHelper console, long executionStartTime) throws IOException, TezException {
+    status = dagClient.getDAGStatus(new HashSet<StatusGetOpts>());
```

Comment:
updated status is already available when this progress monitor is created. Can we re-use that to avoid another rpc ? Maybe pass as argument to updateFUnction ?